deepgram/errors.py

```diff
@@ -0,0 +1,87 @@
+from typing import Union, Optional
+import websockets.exceptions
+import urllib.error
+import aiohttp
+import uuid
+
+
+class DeepgramError(Exception):
+    pass
+
+
+class DeepgramSetupError(DeepgramError, ValueError):
+    pass
+
+
+class DeepgramApiError(DeepgramError):
+    """An error returned by the Deepgram API.
+
+    If the error was raised by an http client, the client's error message
+    is accessible via the `http_library_error` field. This may be useful
+    for handling different error codes, such as 429s or 503s.
+
+    The `error` field is set to the API's error message (dict), if avilable.
+    Otherwise the `error` field is set to the parent exception's message (str).
+    
+    The `warning` field is set to the API's warning messages (list[str]), if available.
+
+    The `request_id` field is set to the API's request ID, if available.
+    """
+    def __init__(
+        self,
+        *args: object,
+        http_library_error: Optional[
+            Union[
+                urllib.error.HTTPError,
+                urllib.error.URLError,
+                websockets.exceptions.InvalidHandshake,
+                aiohttp.ClientResponseError,
```

Comment:
Does this mean we are presenting all of these errors as a type of Deepgram error? Would we maybe want other non-DG errors to bubble up rather than wrapping them as our own?

Reply:
In the previous iteration of the code, the Python SDK was catching all non-DG errors and returning them as `Exception`, sometimes without using the `from` operator (!!!). That means the non-DG errors - and the information available in them - were inaccessible.

I agree that we could allow these errors to bubble up on their own rather than catch them and wrap them in a Deepgram error. It would allow users to be more explicit about how to handle errors, but it would also make error handling more difficult for the user because they would need to handle multiple exception types.

The two approaches are targeted at different audiences: The implementation where errors are wrapped up into a single DG exception makes it easy to do simply error handling and therefore makes it easy for newer programmers to handle failed API requests. On the other hand, the implementation where errors bubble up requires the user to be explicit about how they handle the different use cases, which is likely more useful for power users.

The code I wrote grabs the underlying non-DG error and adds it as a field to the DG error. I hope this satisfies both user groups because: the programmers with less experience can simply catch one error, and the power users can grab the underlying HTTP error and perform more advanced error handling.

Aside: I said, "the implementation where errors bubble up requires the user to be explicit about how they handle the different use cases". That isn't completely true because users could always write `except Exception:` to catch _all_ errors, but I try to highly discourage the use of `except Exception`. It is almost always too broad. For example, if a user passes an incorrect type that results in a `TypeError`, they wouldn't want their retry logic to trigger and keep making API calls with the same `TypeError`.

After that long explanation, I hope this code serves both new users and power users by making it easy to handle failed API requests in both simple and complex ways.